From 4a41b18ccc43ea865cd3fa2f3920d9d5b22e7bb0 Mon Sep 17 00:00:00 2001
From: Sridharan S N <quic_sridsn@quicinc.com>
Date: Wed, 1 Nov 2023 16:45:29 +0530
Subject: [PATCH] mtd: nand: Add winbond W25N512GW, W25N04KWZEIR and
 W25N02KWZEIR devices

This patch adds the winbond W25N512GW and W25N02KWZEIR devices
and fixes the W25N01GWZEIG device id from 0xBA to 0xBA21,
W25N02JWZEIF device id from 0xBF to 0xBF22 and
W25N01JW device id from 0xBC to 0xBC21

Change-Id: I58819d2a500b3d5d1108c772e444ce4afacef48f
Signed-off-by: Sridharan S N <quic_sridsn@quicinc.com>
---
 drivers/mtd/nand/raw/nand_ids.c   | 24 ++++++++++++++++++------
 drivers/mtd/nand/raw/qcom_nandc.c |  8 ++++++++
 2 files changed, 26 insertions(+), 6 deletions(-)

diff --git a/drivers/mtd/nand/raw/nand_ids.c b/drivers/mtd/nand/raw/nand_ids.c
index ebdd192cddd1..432596e31da0 100644
--- a/drivers/mtd/nand/raw/nand_ids.c
+++ b/drivers/mtd/nand/raw/nand_ids.c
@@ -85,8 +85,8 @@ struct nand_flash_dev nand_flash_ids[] = {
 		{ .id = {0xc8, 0x21} },
 		SZ_2K, SZ_128, SZ_128K, 0, 2, 64, NAND_ECC_INFO(4, SZ_512)},
 	{"W25N01JW SPI NAND 1.8V 1G-BIT",
-		{ .id = {0xef, 0xbc} },
-		SZ_2K, SZ_128, SZ_128K, 0, 2, 64, NAND_ECC_INFO(4, SZ_512)},
+		{ .id = {0xef, 0xbc, 0x21} },
+		SZ_2K, SZ_128, SZ_128K, 0, 3, 64, NAND_ECC_INFO(4, SZ_512)},
 	{"GD5F1GQ5REYIG SPI NAND 1G",
 		{ .id = {0xc8, 0x41} },
 		SZ_2K, SZ_128, SZ_128K, 0, 2, 128, NAND_ECC_INFO(8, SZ_512)},
@@ -100,8 +100,8 @@ struct nand_flash_dev nand_flash_ids[] = {
 		{ .id = {0xc8, 0x25} },
 		SZ_2K, SZ_512, SZ_128K, 0, 2, 64, NAND_ECC_INFO(4, SZ_512)},
 	{"W25N02JWZEIF SPI NAND 2G 1.8V",
-		{ .id = {0xef, 0xbf} },
-		SZ_2K, SZ_256, SZ_128K, 0, 2, 64, NAND_ECC_INFO(4, SZ_512)},
+		{ .id = {0xef, 0xbf, 0x22} },
+		SZ_2K, SZ_256, SZ_128K, 0, 3, 64, NAND_ECC_INFO(4, SZ_512)},
 	{"MX35UF1GE4AC SPI NAND 1G 1.8V",
 		{ .id = {0xc2, 0x92} },
 		SZ_2K, SZ_128, SZ_128K, 0, 2, 64, NAND_ECC_INFO(4, SZ_512)},
@@ -116,8 +116,20 @@ struct nand_flash_dev nand_flash_ids[] = {
 		SZ_2K, SZ_256, SZ_128K, 0, 2, 128, NAND_ECC_INFO(8, SZ_512)},
 
 	{"W25N01GWZEIG SPI NAND 1G 1.8V",
-		{ .id = {0xef, 0xba} },
-		SZ_2K, SZ_128, SZ_128K, 0, 2, 64, NAND_ECC_INFO(4, SZ_512)},
+		{ .id = {0xef, 0xba, 0x21} },
+		SZ_2K, SZ_128, SZ_128K, 0, 3, 64, NAND_ECC_INFO(4, SZ_512)},
+
+	{"W25N02KWZEIR SPI NAND 2G 1.8V",
+		{ .id = {0xef, 0xba, 0x22} },
+		SZ_2K, SZ_256, SZ_128K, 0, 3, 128, NAND_ECC_INFO(8, SZ_512)},
+
+	{"W25N04KWZEIR SPI NAND 4G 1.8V",
+		{ .id = {0xef, 0xba, 0x23} },
+		SZ_2K, SZ_512, SZ_128K, 0, 3, 128, NAND_ECC_INFO(8, SZ_512)},
+
+	{"W25N512GW SPI NAND 512M 1.8V",
+		{ .id = {0xef, 0xba, 0x20} },
+		SZ_2K, SZ_64, SZ_128K, 0, 3, 64, NAND_ECC_INFO(4, SZ_512)},
 
 	LEGACY_ID_NAND("NAND 4MiB 5V 8-bit",   0x6B, 4, SZ_8K, SP_OPTIONS),
 	LEGACY_ID_NAND("NAND 4MiB 3,3V 8-bit", 0xE3, 4, SZ_8K, SP_OPTIONS),
diff --git a/drivers/mtd/nand/raw/qcom_nandc.c b/drivers/mtd/nand/raw/qcom_nandc.c
index 52cc44bcfe23..85b311aa906c 100644
--- a/drivers/mtd/nand/raw/qcom_nandc.c
+++ b/drivers/mtd/nand/raw/qcom_nandc.c
@@ -308,6 +308,8 @@ nandc_set_reg(chip, reg,			\
  */
 #define NAND_ERASED_CW_SET		BIT(4)
 
+#define NAND_MID_WINBOND		0xEF
+
 /*
  * An array holding the fixed pattern
  */
@@ -1604,6 +1606,12 @@ static int read_id(struct qcom_nand_host *host, int column)
 
 	read_reg_dma(nandc, NAND_READ_ID, 1, NAND_BAM_NEXT_SGL);
 
+	if (nandc->props->is_serial_nand &&
+		((le32_to_cpu(nandc->reg_read_buf[0]) & 0xFF) ==
+		NAND_MID_WINBOND)) {
+		nandc->buf_count = 4;
+	}
+
 	return 0;
 }
 
-- 
2.34.1

